vernac/declaremods.ml

```diff
@@ -537,77 +657,204 @@ let rec get_module_sobjs is_mod env inl = function
+  in
+  let node = ModuleObject (id,sobjs) in
+  (* We add the keep objects, if any, and if this isn't a functor *)
+  let objects = match keep, m_info.cur_mbids with
+    | [], _ | _, _ :: _ -> special@[node]
+    | _ -> special@[node;KeepObject (id,keep)]
+  in
+  (* Name consistency check : start_ vs. end_module *)
+  (*
+  Printf.eprintf "newoname=%s, oldoname=%s\n" (string_of_path (fst newoname)) (string_of_path (fst oldoname));
+  assert (DirPath.equal (Lib.prefix()).Nametab.obj_dir olddp);
+  assert (ModPath.equal oldprefix.Nametab.obj_mp mp);
+  *)
+  (* Printf.eprintf "newoname=%s, oldoname=%s\n" (string_of_path (fst newoname)) (string_of_path (fst oldoname)); *)
+  (* Printf.eprintf "newoname=%s, cur_mp=%s\n" (ModPath.debug_to_string (mp_of_kn (snd newoname))) (ModPath.debug_to_string m_info.cur_mp); *)
+
+  m_info.cur_mp, objects
+
+let end_module () =
+  let oldprefix,fs,objects = Lib.Synterp.end_module () in
+  let m_info = !openmod_syntax_info in
+  let olddp, id = split_dirpath oldprefix.Nametab.obj_dir in
+  let mp,objects = end_module_core id m_info objects fs in
+
+  let () = SynterpVisitor.add_leaves objects in
+
+  (* Name consistency check : kernel vs. library *)
+  (* CDebug.debug_synterp (fun () -> Pp.(str"prefix=" ++ DirPath.print ((Lib.prefix()).Nametab.obj_dir) ++ str", olddp=" ++ DirPath.print olddp)); *)
+  assert (DirPath.equal (Lib.prefix()).Nametab.obj_dir olddp);
+  mp
+
+let get_functor_sobjs is_mod inl (mbids,mexpr) =
+  let (mbids0, aobjs) = SynterpVisitor.get_module_sobjs is_mod () inl mexpr in
+  (mbids @ mbids0, aobjs)
+
+let declare_module id args res mexpr_o fs =
+  (* We simulate the beginning of an interactive module,
+     then we adds the module parameters to the global env. *)
+  let mp = ModPath.MPdot(!openmod_syntax_info.cur_mp, Label.of_id id) in
+  let args = intern_args args in
+  let mbids = List.flatten @@ List.map fst args in
+  let mty_entry_o = match res with
+    | Enforce (mty,ann) ->
+        let inl = inl2intopt ann in
+        let (mte, base, kind) = Modintern.intern_module_ast Modintern.ModType mty in
+        Enforce (mte, base, kind, inl)
+    | Check mtys ->
+      Check (build_subtypes mtys)
+  in
+  let mexpr_entry_o = match mexpr_o with
+    | None -> None
+    | Some (mexpr,ann) ->
+      let (mte, base, kind) = Modintern.intern_module_ast Modintern.Module mexpr in
+      Some (mte, base, kind, inl2intopt ann)
+  in
+  let sobjs, mp0 = match mexpr_entry_o, mty_entry_o with
+    | None, Check _ -> assert false (* No body, no type ... *)
+    | _, Enforce (typ,_,_,inl_res) -> get_functor_sobjs false inl_res (mbids,typ), get_module_path typ
+    | Some (body, _, _, inl_expr), Check _ ->
+      get_functor_sobjs true inl_expr (mbids,body), get_module_path body
+  in
+  (* Undo the simulated interactive building of the module
+     and declare the module as a whole *)
+  Summary.unfreeze_summaries ~partial:true fs; (* TODO check this *)
```

Comment:
Is this TODO intended to be solved before merging?

Reply:
No, it was actually the same as @gares ' question. I removed the TODO.